### PR TITLE
Deprecate custom `AllGatherGrad` implementation

### LINKED ISSUE #10445: Remove custom `AllGatherGrad` implementation in favor of `torch.distributed`'s
### Motivation

Lightning has a utility defined for all gather with gradients here: https://github.com/PyTorchLightning/pytorch-lightning/blob/d515bcac969c2a485ada673e302bfac51f142331/pytorch_lightning/utilities/distributed.py#L200-L222

However, this is already available in torch distributed: https://github.com/pytorch/pytorch/blob/6b44e75f6bccca7acc8ec31a635f1175c265ac54/torch/distributed/nn/functional.py#L82-L94

So there's no need to redefine this in Lightning

### Pitch

Remove the custom all gather grad implementation and call torch distributed's functional API 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @rohitgr7 @akihironitta @justusschock @tchaton

### PR DESCRIPTION
**Pull request by puhuk**
Remove custom AllGatherGrad implementation

## What does this PR do?

Remove custom AllGatherGrad implementation in favor of torch.distributed's

Fixes #10445

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -112,6 +112,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Removed deprecated `utilities.distributed.rank_zero_{warn/deprecation}` ([#10451](https://github.com/PyTorchLightning/pytorch-lightning/pull/10451))
 
+- Removed deprecated custom `AllGatherGrad implementation` ([#10504](https://github.com/PyTorchLightning/pytorch-lightning/pull/10504))
```
Comment (carmocca) [inline CHANGELOG.md:115]:
```suggestion

- Removed custom `AllGatherGrad` implementation in favor of PyTorch's ([#10504](https://github.com/PyTorchLightning/pytorch-lightning/pull/10504))
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -177,31 +177,6 @@ def sync_ddp(
...
-
-    @staticmethod
-    def backward(ctx: Any, *grad_output: torch.Tensor) -> Tuple[torch.Tensor, None]:
-        grad_output = torch.cat(grad_output)
-
-        torch.distributed.all_reduce(grad_output, op=torch.distributed.ReduceOp.SUM, async_op=False, group=ctx.group)
-
-        return grad_output[torch.distributed.get_rank()], None
-
-
```
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py:204]: this class can't be immediately removed like this as it is part of the stable, public Lightning API. https://pytorch-lightning.readthedocs.io/en/latest/governance.html#api-evolution

Instead, this class should be marked as deprecated in version 1.6 and slated for removal in version 1.8.

However, the implementation of `all_gather_ddp_if_available` can be changed freely to rely on `torch.distributed.all_gather`

**Review by ananthsub — CHANGES_REQUESTED** [review]
thanks for picking this up @puhuk !

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (puhuk) [reply pytorch_lightning/utilities/distributed.py:204]: Thanks for review! Let me check and resend PR :)

File: CHANGELOG.md [hunk 3]
```diff
@@ -113,6 +113,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed deprecated `utilities.distributed.rank_zero_{warn/deprecation}` ([#10451](https://github.com/PyTorchLightning/pytorch-lightning/pull/10451))
 
 
+- Removed custom `AllGatherGrad` implementation in favor of PyTorch's ([#10504](https://github.com/PyTorchLightning/pytorch-lightning/pull/10504))
```
Comment (akihironitta) [inline CHANGELOG.md:116]: Could you move this line to "Deprecated" section above and follow ananthsub's comment?

Comment (puhuk) [reply CHANGELOG.md:116]: Ok, let me change this.

File: tests/utilities/test_all_gather_grad.py [hunk 4]
```diff
@@ -19,34 +18,6 @@ def setup_ddp(rank, world_size):
         torch.distributed.init_process_group("gloo", rank=rank, world_size=world_size)
 
 
-def _test_all_gather_ddp(rank, world_size):
```
Comment (tchaton) [inline tests/utilities/test_all_gather_grad.py:22]: We should keep this test until v1.6 as we are depreciating `AllGatherGrad `

**Comment by puhuk** [comment]
I think error comes because all_gather from pytorch and lightning is a bit different.
Pytorch torch.distributed.all_gather : takes tensor_list, tensor, group as input
Lightning all_gather : takes data, group and sync_grads as input

Is that all right to use torch.distributed.all_gather instead of lightning all_gather?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 5]
```diff
@@ -455,7 +455,8 @@ def _share_information_to_prevent_deadlock(self) -> None:
     def _share_pids(self) -> None:
         """Make all DDP processes aware of all processes pids."""
         self.barrier()
-        pids = self.all_gather(torch.tensor(os.getpid(), device=self.root_device))
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp.py:458]: Let's rely on self.all_gather instead.

File: CHANGELOG.md [hunk 6]
```diff
@@ -130,25 +130,18 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 
 
-- Removed deprecated `disable_validation` property from Trainer ([#10450](https://github.com/PyTorchLightning/pytorch-lightning/pull/10450))
-
-
-- Removed deprecated `CheckpointConnector.hpc_load` property in favor of `CheckpointConnector.restore` ([#10525](https://github.com/PyTorchLightning/pytorch-lightning/pull/10525))
-
-
-- Removed deprecated `reload_dataloaders_every_epoch` from `Trainer` in favour of `reload_dataloaders_every_n_epochs` ([#10481](https://github.com/PyTorchLightning/pytorch-lightning/pull/10481))
+- Removed deprecated `Trainer.train_loop` property in favor of `Trainer.fit_loop` ([#10482](https://github.com/PyTorchLightning/pytorch-lightning/pull/10482))
```
Comment (tchaton) [inline CHANGELOG.md:142]: bad merge ?

File: CHANGELOG.md [hunk 7]
```diff
@@ -130,25 +130,18 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed deprecated `utilities.distributed.rank_zero_{warn/deprecation}` ([#10451](https://github.com/PyTorchLightning/pytorch-lightning/pull/10451))
 
 
-- Removed deprecated `mode` argument from `ModelSummary` class ([#10449](https://github.com/PyTorchLightning/pytorch-lightning/pull/10449))
+- Removed custom `AllGatherGrad` implementation in favor of PyTorch's ([#10504](https://github.com/PyTorchLightning/pytorch-lightning/pull/10504))
 
 
-- Removed deprecated `Trainer.train_loop` property in favor of `Trainer.fit_loop` ([#10482](https://github.com/PyTorchLightning/pytorch-lightning/pull/10482))
+- Removed deprecated `mode` argument from `ModelSummary` class ([#10449](https://github.com/PyTorchLightning/pytorch-lightning/pull/10449))
```
Comment (tchaton) [inline CHANGELOG.md:136]: Bad merge ?

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 5 above)
Comment (puhuk) [reply pytorch_lightning/plugins/training_type/ddp.py:458]: May I ask what `rely on` mean? You mean just using self.all_gather instead of pytorch's all_gather?

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp.py:458]: Yes

**Comment by puhuk** [comment]
May I ask review for this PR

**Comment by awaelchli** [comment]
Hello @puhuk 
I decided to help a bit here because a few things were missing and it appears from the commit history there were some struggles here to get it working. At the very beginning of the PR, you were actually on the right path already ❤️ 
So let's see if my fixes did something good here :) 

On another note, I noticed that the old AllGatherGrad had backward() implemented. It is not exactly clear to me how `torch.distributed.all_gather` is supposed to be a replacement for this. I will comment on the linked issue on this matter.

**Comment by puhuk** [comment]
@awaelchli 
Really thanks for your help.
Let me study with your code and commit :)
Thanks again.
